doc/api/http2.md

```diff
@@ -34,34 +45,34 @@ server.on('stream', (stream, headers) => {
   stream.end('<h1>Hello World</h1>');
 });
 
-server.listen(80);
+server.listen(8443);
 ```
 
-Note that the above example is an HTTP/2 server that does not support SSL.
-This is significant as most browsers support HTTP/2 only with SSL.
-To make the above server be able to serve content to browsers,
-replace `http2.createServer()` with
-`http2.createSecureServer({key: /* your SSL key */, cert: /* your SSL cert */})`.
+### Client-side example
 
 The following illustrates an HTTP/2 client:
 
 ```js
-const http2 = require('http2');
-
-const client = http2.connect('http://localhost:80');
+var http2 = require('http2');
+const client = http2.connect('https://localhost');
+client.on('socketError', (err) => console.error(err))
+client.on('error', (err) => console.error(err))
 
-// req is a Duplex
 const req = client.request({ ':path': '/' });
 
-req.on('response', (headers) => {
-  console.log(headers[':status']);
-  console.log(headers['date']);
+req.on('response', (headers, flags) => {
+  for (var name in headers) {
+    console.log(name + ": " + headers[name]);
+  }
 });
 
-let data = '';
-req.setEncoding('utf8');
-req.on('data', (d) => data += d);
-req.on('end', () => client.destroy());
+let data = []
+req.on('data', (d) => data.push(d));
+req.on('end', () => {
+  console.log();
+  console.log(Buffer.concat(data).toString('utf8'));
```

Comment:
I feel like the example with `setEncoding` might've been a bit clearer but I'm ok with either.

Reply:
The problem with `setEncoding` is that it turns each chunk into a string, meaning Buffer.concat won't work. We could join strings with Array.prototype.join, but to me, accumulating chunks in buffers is more natural. It's also more compatible with performing other transforms, like inflate or gunzip, on the resulting data.

Reply:
(Making it more concrete: I actually spent a while fighting a subtle bug with this, trying to inflate data that had been decoded as if it were utf-8)

Reply:
I kind of agree with @apapirovski here – you’re right, if you don’t want to receive text data, you shouldn’t use `setEncoding()` at all, but `Buffer.concat()` has the downside of having double the memory requirement, because the concatenated buffer needs to be a fresh allocation too…

Reply:
My understanding was that doing `+=` on a string in a for loop has bad performance: it potentially causes an allocation each time plus O(n) work to copy into the newly allocated space. Does Node specially detect and optimize for this case, e.g. by predicting the final length of the string and preallocating for it?

Reply:
Been a while since I read about it but pretty sure V8 has a special way of dealing with string concatenation. I believe it only gets truly concatenated when it actually gets used so if you're not also reading (or otherwise using) the string within your for loop then it'll always outperform arrays.

But anyway, it feels like this discussion is heading in the wrong direction. The main concern here should be the ease of understanding for somebody new to Node. To me, the `setEncoding` example is the only one that doesn't rely on understanding yet another part of Node and requires less lines to get the same point across.

Reply:
> I believe it only gets truly concatenated when it actually gets used so if you're not also reading (or otherwise using) the string within your for loop then it'll always outperform arrays.

Aha, I did not know that! Thanks for updating my knowledge. I also checked the `http` docs, and they use the string concatenation style, so I've reverted to that.

> The main concern here should be the ease of understanding for somebody new to Node.

True, although since example code so often forms the kernel of a new program, it's important not to recommend anti-patterns that a newbie might not know enough to see. Of course, in this case it turned out to not be an anti-pattern in modern environments, which is great!